python-api/tests/integrationtests/test_controller.py

```diff
@@ -3,725 +3,343 @@
+        self.setup_server()
+        self.setup_controller()
 
-            serv.terminate(1)
-        finally:
-            serv.terminate_and_output_status(cov=True)
+        self.log.info("starting 1 test-audio source")
+        self.new_test_audio()
 
+        self.wait_until_ready()
 
-class VideoFileSink(object):
+        self.log.info("calling get_audio_port")
+        assert self.controller.get_audio_port() == 3003
 
-    """Sink the video to a file
+
+class TestGetPreviewPorts(IntegrationTestbase):
+    """ Test get_preview_ports method
     """
+    def wait_for_sources(self, count):
+        """ Blocks until the Server has reported, that the right number of
+        preview-ports are is started
+        """
+        self.log.info("waiting for Server to start preview-port-outputs")
+        self.serv.wait_for_output('tcpserversink name=sink', count=count)
+
+    def test_one_preview_port(self):
+        """Test get_preview_ports method returning the expected port"""
+        self.setup_server()
+        self.setup_controller()
+
+        self.log.info("starting 1 test-video source")
+        self.new_test_video()
+
+        self.wait_for_sources(count=1)
+
+        self.log.info("calling get_preview_ports")
+        assert self.controller.get_preview_ports() == [3003]
+
+    def test_five_preview_ports(self):
+        """Test get_preview_ports method returning the expected port"""
+        self.setup_server()
+        self.setup_controller()
+
+        self.log.info("starting 5 test-video sources")
+        self.new_test_video()
+        self.new_test_video()
+        self.new_test_video()
+        self.new_test_video()
+        self.new_test_video()
+
+        self.log.info("waiting for the test-video sources to come up")
+        self.wait_for_sources(count=5)
+
+        self.log.info("calling get_preview_ports")
+        assert self.controller.get_preview_ports() == [
+            3003, 3004, 3005, 3006, 3007]
+
+
+class TestSignals(IntegrationTestbaseMainloop):
+    """ Test the various on_* signal-handler methods
+    """
+    def wait_for_sources(self, count):
+        """ Blocks until the Server has reported, that the right number of
+        preview-ports are is started
+        """
+        self.log.info("waiting for Server to start preview-port-outputs")
+        self.serv.wait_for_output('tcpserversink name=sink', count=count)
+
+    def setup_sources(self):
+        """ Starts some Test-Video streams and waits until they are ready
+        """
+        self.log.info("starting 2 test-video sources")
+        self.new_test_video()
+        self.new_test_video()
+
+        self.log.info("waiting for the test-video sources to come up")
+        self.wait_for_sources(count=2)
+
+    def test_initial_mode_callback(self):
+        """Create a Controller object, call on_new_mode_online method and
+        check that the callback fires initially when the sources are set up
+        """
```

Comment:
I'm a little confused what is going on in this test? Is it testing that after a second sources is added, the controller gets a callback that the composite mode has changed?